govuk-one-login/authentication-api
# BAU: Add detect-secrets pre-commit hook

## What

This will scan for potential secrets inadvertently added to code.

`.secrets.baseline` contains an initial baseline of all existing
'violations' in the repository, so we don't immediately have hook
failures. These are all false positives.

A lot of these false positives are coming from the java tests. To avoid
creating new detections in future, there are a few options:

1. Move all the test passwords to a single file in `shared` and import
   them in the tests. Using these variables will stop the detections.
2. Add the 'ignore secret' comments when something is detected: https://github.com/Yelp/detect-secrets/tree/master?tab=readme-ov-file#inline-allowlisting

## How to review

1. Ensure the pre-commit action on this PR has completed successfully
2. Check out the branch
3. Create a file with a 'secret':

```sh
#!/usr/bin/env bash

PASSWORD="test-pw"
TEST="f572d396fae9206628714fb2ce00f72e94f2258f"
```

4. ensure [pre-commit is set up](https://pre-commit.com/#install)
5. install it in your local checkout: `pre-commit install`
6. Attempt to commit the file you created: `git add $file; git commit`
7. Ensure pre-commit reports an error:

```
Detect secrets...........................................................Failed
- hook id: detect-secrets
- exit code: 1

ERROR: Potential secrets about to be committed to git repo!

Secret Type: Secret Keyword
Location:    test.sh:3

Secret Type: Hex High Entropy String
Location:    test.sh:4
```
